### PR TITLE
[RHPAM-3640] Configure Kafka JBPM Emitter from Operator

### PR DESCRIPTION
**Pull request by desmax74**
See:https://issues.redhat.com/browse/RHPAM-3640
7.12.0
Signed-off-by: desmax74 <mdessi@redhat.com>

@spolti @jakubschwan

### CODE REVIEW

**Review by tchughesiv — CHANGES_REQUESTED** [review]
this PR should probably wait until https://github.com/kiegroup/kie-cloud-operator/pull/654 is merged. 
this PR should probably cherry-pick commits from the other PR and then build on top of it with 7.12 related changes. this is a critical step so that the `config/7.11.1` files from prior releases match in a future 7.12.0 release. This is important because we support 2 product versions per operator release.

File: deploy/crds/kieapp.crd.yaml
Comment (jakubschwan) [inline deploy/crds/kieapp.crd.yaml:1916]: There are some configuration endpoints missing that may be handy for users.
As I can see we are missing here property for date format configuration and properties for topics (processes, tasks, cases). See [KafkaEventEmitter](https://github.com/kiegroup/jbpm/blob/master/jbpm-event-emitters/jbpm-event-emitters-kafka/src/main/java/org/jbpm/event/emitters/kafka/KafkaEventEmitter.java)
They are not required to be set.

File: pkg/controller/kieapp/defaults/defaults.go
Comment (jakubschwan) [inline pkg/controller/kieapp/defaults/defaults.go:1086]: We don't need to set this default by ourself, this is already part of KafkaEventEmitter.java. I think it's enough to have it mentioned just in parameter description in KieApp.yaml file.

File: deploy/crds/kieapp.crd.yaml
Comment (jakubschwan) [inline deploy/crds/kieapp.crd.yaml:1924]: There is no default value set for the property.
```suggestion
                                before considering a request complete.
```

File: deploy/crds/kieapp.crd.yaml
Comment (jakubschwan) [inline deploy/crds/kieapp.crd.yaml:1938]: missing default value

File: pkg/controller/kieapp/defaults/defaults.go
Comment (desmax74) [reply pkg/controller/kieapp/defaults/defaults.go:1086]: Ok

File: deploy/crds/kieapp.crd.yaml
Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml:1938]: ok

File: deploy/crds/kieapp.crd.yaml
Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml:1924]: ok

File: deploy/crds/kieapp.crd.yaml
Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml:1916]: Added date format

File: pkg/controller/kieapp/defaults/upgrade_test.go
Comment (tchughesiv) [inline pkg/controller/kieapp/defaults/upgrade_test.go:98]: these upgrade test changes may be unnecessary in this branch because you still need to cherry-pick the commit from the other 7.11.1 PR so that those 7.11.1 configs make it into the main branch as well. once that cherry-picking occurs... there may not be any diffs between version configs.

**Review by tchughesiv — CHANGES_REQUESTED** [review]

File: pkg/controller/kieapp/defaults/upgrade_test.go
Comment (desmax74) [reply pkg/controller/kieapp/defaults/upgrade_test.go]: ok, reverted

**Comment by tchughesiv** [comment]
/hold
@desmax74 still not seeing the 7.11.1 config changes, from https://github.com/kiegroup/kie-cloud-operator/pull/654, in this PR

File: pkg/controller/kieapp/defaults/defaults_test.go
Comment (jakubschwan) [inline pkg/controller/kieapp/defaults/defaults_test.go:5955]: RHDM environment with Kafka JBPM configuration does not make sense, as RHDM Kie Server do not have enabled JBPM capabilities. For RHDM it would be better to not have this configuration (but this I'm worried is not possible to have in KieApp) or have parameter KIE_SERVER_KAFKA_EXT_ENABLED override to false with some exception.
Please tests this configuration on RHPAM environment.

File: deploy/crds/kieapp.crd.yaml
Comment (jakubschwan) [reply deploy/crds/kieapp.crd.yaml:1916]: Still missing topic mapping properties
org.kie.jbpm.event.emitters.kafka.topic.processes
org.kie.jbpm.event.emitters.kafka.topic.cases
org.kie.jbpm.event.emitters.kafka.topic.tasks

Comment (spolti) [reply deploy/crds/kieapp.crd.yaml:1916]: these are added by the image, seems they are hardcoded, Max, is this the expected behaviour?

File: pkg/apis/app/v2/kieapp_types.go
Comment (spolti) [inline pkg/apis/app/v2/kieapp_types.go:993]: isn't missing a this property?
https://github.com/jboss-container-images/jboss-kie-modules/pull/456/files#diff-3e7628a0c0671f1c47f2f4da279754577748d791191318fb85c3d250b1ca182cR42

File: pkg/controller/kieapp/defaults/defaults_test.go
Comment (spolti) [reply pkg/controller/kieapp/defaults/defaults_test.go:5955]: +1

**Review by spolti — CHANGES_REQUESTED** [review]

File: deploy/crds/kieapp.crd.yaml
Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml:1916]: Yes, are hard coded in the emitter's code ithis call https://github.com/kiegroup/jbpm/blob/master/jbpm-event-emitters/jbpm-event-emitters-kafka/src/main/java/org/jbpm/event/emitters/kafka/KafkaEventEmitter.java#L163 and here https://github.com/kiegroup/jbpm/blob/master/jbpm-event-emitters/jbpm-event-emitters-kafka/src/main/java/org/jbpm/event/emitters/kafka/KafkaEventEmitter.java#L91..L106

File: pkg/controller/kieapp/defaults/defaults_test.go
Comment (desmax74) [reply pkg/controller/kieapp/defaults/defaults_test.go]: ok

File: pkg/apis/app/v2/kieapp_types.go
Comment (desmax74) [reply pkg/apis/app/v2/kieapp_types.go:993]: No, we add in the common.yaml if the KafkaJbpmEventEmitters is present https://github.com/kiegroup/kie-cloud-operator/pull/654/files#diff-66b9e589dd9cc4038e91fe574a84c9d024cc6e65956366412f929a815e43cebaR946

Comment (spolti) [reply pkg/apis/app/v2/kieapp_types.go:993]: ok

**Comment by spolti** [comment]
> this PR should probably wait until #654 is merged.
> this PR should probably cherry-pick commits from the other PR and then build on top of it with 7.12 related changes. this is a critical step so that the `config/7.11.1` files from prior releases match in a future 7.12.0 release. This is important because we support 2 product versions per operator release.

Thanks for the heads up.

**Comment by tchughesiv** [comment]
> > this PR should probably wait until #654 is merged.
> > this PR should probably cherry-pick commits from the other PR and then build on top of it with 7.12 related changes. this is a critical step so that the `config/7.11.1` files from prior releases match in a future 7.12.0 release. This is important because we support 2 product versions per operator release.
> 
> Thanks for the heads up.

might be worth considering no longer supporting prior product product versions within a single operator release. this can really only be removed, IMO, once dedicated version channels are added. Maybe even at patch level instead?
https://issues.redhat.com/browse/KIECLOUD-372

**Comment by spolti** [comment]
@desmax74 once we merge https://github.com/kiegroup/kie-cloud-operator/pull/654, you can go ahead and cherry-pick the 7.11.x commit as Tommy explained.

**Comment by desmax74** [comment]
close, replaced with https://github.com/kiegroup/kie-cloud-operator/pull/667

**Pull request closed without merging**
